File: src/2025-02-24-deepclaude-rust/python/exercises/practice/zipper/zipper.py
class Zipper:
    def __init__(self, focus, crumbs):
        self.focus = focus
        self.crumbs = crumbs

    @staticmethod
    def from_tree(tree):
        return Zipper(tree, [])

    def value(self):
        return self.focus['value']

    def set_value(self, value):
        new_focus = {'value': value, 'left': self.focus.get('left'), 'right': self.focus.get('right')}
        return Zipper(new_focus, self.crumbs)

    def left(self):
        if not self.focus.get('left'):
            return None
        new_crumbs = [{'value': self.focus['value'], 'right': self.focus.get('right'), 'dir': 'left'}] + self.crumbs
        return Zipper(self.focus['left'], new_crumbs)

    def set_left(self, left):
        new_focus = {'value': self.focus['value'], 'left': left, 'right': self.focus.get('right')}
        return Zipper(new_focus, self.crumbs)

    def right(self):
        if not self.focus.get('right'):
            return None
        new_crumbs = [{'value': self.focus['value'], 'left': self.focus.get('left'), 'dir': 'right'}] + self.crumbs
        return Zipper(self.focus['right'], new_crumbs)

    def set_right(self, right):
        new_focus = {'value': self.focus['value'], 'left': self.focus.get('left'), 'right': right}
        return Zipper(new_focus, self.crumbs)

    def up(self):
        if not self.crumbs:
            return None
        crumb = self.crumbs[0]
        if crumb['dir'] == 'left':
            new_focus = {'value': crumb['value'], 'left': self.focus, 'right': crumb['right']}
        else:
            new_focus = {'value': crumb['value'], 'left': crumb['left'], 'right': self.focus}
        return Zipper(new_focus, self.crumbs[1:])

    def to_tree(self):
        current = self
        while current.crumbs:
            current = current.up()
        return current.focus

File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/zipper/zipper.js
//
// This is only a SKELETON file for the 'Zipper' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Zipper {
  constructor(focus, crumbs) {
    this.focus = focus;
    this.crumbs = crumbs;
  }

  static fromTree(tree) {
    return new Zipper(tree, []);
  }

  toTree() {
    let current = this;
    while (current.crumbs.length > 0) {
      current = current.up();
      if (!current) break;
    }
    return current ? current.focus : null;
  }

  value() {
    return this.focus.value;
  }

  left() {
    if (!this.focus.left) return null;
    const crumb = {
      dir: 'left',
      value: this.focus.value,
      right: this.focus.right,
    };
    return new Zipper(this.focus.left, [...this.crumbs, crumb]);
  }

  right() {
    if (!this.focus.right) return null;
    const crumb = {
      dir: 'right',
      value: this.focus.value,
      left: this.focus.left,
    };
    return new Zipper(this.focus.right, [...this.crumbs, crumb]);
  }

  up() {
    if (this.crumbs.length === 0) return null;
    const crumb = this.crumbs[this.crumbs.length - 1];
    const remainingCrumbs = this.crumbs.slice(0, -1);

    let newFocus;
    if (crumb.dir === 'left') {
      newFocus = {
        value: crumb.value,
        left: this.focus,
        right: crumb.right,
      };
    } else {
      newFocus = {
        value: crumb.value,
        left: crumb.left,
        right: this.focus,
      };
    }

    return new Zipper(newFocus, remainingCrumbs);
  }

  setValue(value) {
    const newFocus = {
      value: value,
      left: this.focus.left,
      right: this.focus.right,
    };
    return new Zipper(newFocus, this.crumbs);
  }

  setLeft(left) {
    const newFocus = {
      value: this.focus.value,
      left: left,
      right: this.focus.right,
    };
    return new Zipper(newFocus, this.crumbs);
  }

  setRight(right) {
    const newFocus = {
      value: this.focus.value,
      left: this.focus.left,
      right: right,
    };
    return new Zipper(newFocus, this.crumbs);
  }
}
